src/poll.rs

```diff
@@ -1,60 +1,231 @@
+    }
+}
+
+impl ZmqPoller {
+    /// Attempt to receive a Multipart message from a ZeroMQ socket with buffering.
+    ///
+    /// If there is any message in the buffer, it will be returned right away.
+    /// If not, a batch of messages up to the capacity of the buffer will be read from the socket.
+    ///
+    /// If nothing was received, the read flag is cleared.
+    pub(crate) fn multipart_recv_buffered(
+        &self,
+        cx: &mut Context<'_>,
+        read_buffer: &mut ReceiverBuffer,
+    ) -> Poll<Option<Result<Multipart>>> {
+        if read_buffer.is_empty() {
+            ready!(self.multipart_poll_read_ready(cx))?;
+
+            let mut buffer = Multipart::default();
+            loop {
+                let mut msg = zmq::Message::new();
+                match self.get_socket().recv(&mut msg, zmq::DONTWAIT) {
+                    Ok(_) => {
+                        let more = msg.get_more();
+                        buffer.push_back(msg);
+                        if !more {
+                            read_buffer.push_back(buffer);
+                            if read_buffer.is_full() {
+                                break Poll::Ready(Some(Ok(read_buffer.pop_front().unwrap())));
+                            }
+
+                            buffer = Multipart::default();
+                        }
                     }
-                }
-                return Ok(Async::Ready(()));
-            }
+                    Err(zmq::Error::EAGAIN) => {
+                        if !buffer.is_empty() {
+                            read_buffer.push_back(buffer);
+                        }
+                        self.clear_read_ready(cx, Ready::readable())?;
 
-            //If it's not ready to send yet, then we basically need to hold onto the message and wait
-            Async::NotReady => {
-                return Ok(Async::NotReady);
+                        if read_buffer.is_empty() {
+                            break Poll::Pending;
+                        } else {
+                            break Poll::Ready(Some(Ok(read_buffer.pop_front().unwrap())));
+                        }
+                    }
+                    Err(e) => break Poll::Ready(Some(Err(e.into()))),
+                }
             }
+        } else {
+            Poll::Ready(Some(Ok(read_buffer.pop_front().unwrap())))
         }
     }
 
-    fn recv_message(&self, msg: &mut zmq::Message) -> Poll<(), Error> {
-        let ready = Ready::readable();
+    /// Attempt to receive a Multipart message from a ZeroMQ socket.
+    ///
+    /// Either the whole multipart at once or nothing is received (according to
+    /// [http://zguide.zeromq.org/page:all#Multipart-Messages], when one part of a multipart message
+    /// has been received, all the others are already available as well).
+    ///
+    /// If nothing was received, the read flag is cleared.
+    pub(crate) fn multipart_recv(&self, cx: &mut Context<'_>) -> Poll<Option<Result<Multipart>>> {
```

Comment:
Perhaps remove the Option<> in the return type.  This function only seemingly would only return Some().

Reply:
Right, good catch. ZMQ sockets are by design "endless", so the Stream probably shouldn't signal EOF.